network/src/client.rs

```diff
@@ -104,24 +124,95 @@ impl Client {
         F: FnOnce(Arc<Api>) -> T, {
         let mut extensions = self.extensions.write();
         let timer = self.timer_loop.new_timer();
-        let api = {
+        let (api, channel, rx) = {
             let p2p_channel = self.p2p_channel.clone();
-            Arc::new(ClientApi {
-                extension: RwLock::new(None),
-                p2p_channel,
-                timer,
-            })
+            let (channel, rx) = crossbeam::unbounded();
+            (
+                Arc::new(ClientApi {
+                    name: Default::default(),
+                    need_encryption: Default::default(),
+                    p2p_channel,
+                    timer,
+                    channel: channel.clone().into(),
+                }),
+                channel,
+                rx,
+            )
         };
         let extension = Arc::new(factory(Arc::clone(&api) as Arc<Api>));
         let name = extension.name();
-        *api.extension.write() = Some(Arc::downgrade(&extension) as Weak<NetworkExtension>);
+
+        let cloned_extension = Arc::clone(&extension);
+
+        let extension_initialized: Arc<(Mutex<bool>, Condvar)> = Default::default();
```

Comment:
I think it would be better rather than using tuple.
1. give it a proper type.
```rust
struct Barrier {
   flag: Mutex<bool>,
   condvar: Condvar
}
impl Barrier {
  fn wait(&self) { ... }
  fn trigger(&self) { ... }
}

let init_barrier = Arc::new(InitCondvar {
   flag: Mutex::new(false),
   condvar: Condvar::new(),
});
```

2. ExtensionMessage::Initialize.
```rust
        let join = Some(spawn(move || {
            if let Ok(ExtensionMessgae::Initialize) = rx.recv() {
                cloned_extension.on_initialize();
            } else {
                 unreachable!("Must be initialized first");
            }
            loop {
                crossbeam::select! {
                    recv(rx) -> msg => {
                        match msg {
...
```

Reply:
1. I don't want to create a new structure for the local only and used once type.
If a tuple repeats an item of the same type, it may be better to give it a clear name even though it's used on local only. But they have different type and the type shows the purpose explicitly in this case.

Reply:
And FYI, `Barrier` is not a good name for this kind of structure.
`Barrier` is a well-known synchronization primitive for thread coordination.
It makes threads wait until all participating threads reach some point.

https://en.cppreference.com/w/cpp/experimental/barrier
https://doc.rust-lang.org/std/sync/struct.Barrier.html
https://linux.die.net/man/3/pthread_barrier_wait

Reply:
I changed it to use a channel.
The crossbeam channel guarantees that the receiver receives sent messages after the sender disconnects the channel.